src/transaction.ts

```diff
@@ -44,35 +63,47 @@ export function assembleTransaction(
     minAccountSequenceLedgerGap: raw.minAccountSequenceLedgerGap,
     extraSigners: raw.extraSigners,
   });
-  for (let i = 0; i < raw.operations.length; i++) {
-    const rawOp = raw.operations[i];
-    if ("function" in rawOp) {
-      const sim = simulated[i];
-      if (!sim) {
-        throw new Error("missing simulated operation");
-      }
-      let footprint = sim.footprint ?? rawOp.footprint;
-      if (!(footprint instanceof xdr.LedgerFootprint)) {
-        footprint = xdr.LedgerFootprint.fromXDR(footprint.toString(), "base64");
-      }
-      const auth = (sim.auth ?? rawOp.auth).map((a) =>
-        a instanceof xdr.ContractAuth
-          ? a
-          : xdr.ContractAuth.fromXDR(a.toString(), "base64"),
-      );
-      // TODO: Figure out a cleaner way to clone these operations
-      txn.addOperation(
-        Operation.invokeHostFunction({
-          function: rawOp.function,
-          parameters: rawOp.parameters,
-          footprint,
-          auth,
-        }),
-      );
-    } else {
-      // TODO: Handle this.
-      throw new Error("Unsupported operation type");
-    }
+
+  // apply the pre-built Auth from simulation onto each Tx/Op/HostFunction invocation
+  for (
+    let hostFnIndex = 0;
+    hostFnIndex < simulation.results.length;
+    hostFnIndex++
+  ) {
+    const rawHostFunction: xdr.HostFunction =
+      rawInvokeHostFunctionOp.functions[hostFnIndex];
+    const simHostFunctionResult: SorobanRpc.SimulateHostFunctionResult =
+      simulation.results[hostFnIndex];
+    rawHostFunction.auth(buildContractAuth(simHostFunctionResult.auth));
+    authDecoratedHostFunctions.push(rawHostFunction);
+  }
```

Comment:
a more idiomatic approach would be, I think:

```suggestion
  // apply the pre-built Auth from simulation onto each Tx/Op/HostFunction 
  // invocation
  const authDecoratedHostFunctions = simulation.results.map((simResult, i) => {
    const hostFn: xdr.HostFunction = rawInvokeHostFunctionOp.functions[i];
    hostFn.auth(buildContractAuth(simResult.auth));
    return hostFn;
  });
```

Reply:
oh nice, I didn't know you could get the the index also, yeah I had started with forEach first to do more functional, but still had a var outside holding index which made it clumsy, this is much cleaner, thanks for the pro tip!